lib_eio/net.ml

```diff
@@ -267,3 +269,49 @@ let with_tcp_connect ?(timeout=Time.Timeout.none) ~host ~service t f =
   | exception (Exn.Io _ as ex) ->
     let bt = Printexc.get_raw_backtrace () in
     Exn.reraise_with_context ex bt "connecting to %S:%s" host service
+
+  let accept_fork_domain ~on_error (domain_mgr, domains) (t : #listening_socket) connection_handler () =
+  Switch.run @@ fun sw ->
+  let flow, addr = accept ~sw t in
+  Fun.protect
+    (fun () ->
+      Semaphore.acquire domains ;
+      Domain_manager.run domain_mgr (fun () ->
+        match connection_handler (flow :> stream_socket) addr with
+        | x -> x
+        | exception ex -> on_error (Exn.add_context ex "handling connection from %a" Sockaddr.pp addr)
+      )
+    )
+    ~finally:(fun () -> Semaphore.release domains)
+
+let run_server ?(max_connections=Int.max_int) ?shutdown ?(additional_domains) ?(on_error=raise) listening_socket connection_handler =
+  (if max_connections < 0 then invalid_arg "max_connections");
+  let connections = Semaphore.make max_connections in
+  let shutdown =
+    match shutdown with
+    | Some p -> p
+    | None -> fst (Promise.create ())
+  in
+  let connection_handler flow addr =
+    connection_handler flow addr ;
+    Semaphore.release connections ;
+  in
+  Switch.run @@ fun sw ->
+  let acceptor : unit -> unit =
+    match additional_domains with
+    | Some (domain_mgr, domains) ->
+      (if domains < 1 then invalid_arg "additional_domains");
+      let domains = Semaphore.make domains in
+      accept_fork_domain (domain_mgr, domains) listening_socket ~on_error connection_handler
+    | None -> (fun () -> accept_fork ~sw listening_socket ~on_error connection_handler)
+  in
+  let rec loop () =
+    Fiber.first
+      (fun () ->
+        Semaphore.acquire connections;
```

Comment:
Make sure this is released on error too!

Reply:
Done.